File: repos_estudiantes_ALVAREZ_primera_entrega/ARUQUIPA_VARGAS_ANA_GLORIA/POLIMORFISMO/Ejercicio13/Mp4.py
class MP4:
    def __init__(self, marca: str, capacidadGb: float):
        self.marca = marca
        self.capacidadGb = capacidadGb
        self.canciones = []  
        self.videos = [] 
    
    def __add__(self, cancion: list):
        nombre, artista, peso = cancion
        for c in self.canciones:
            if c[0] == nombre and c[1] == artista:
                print(f"La canción '{nombre}' ya existe.")
                return self

        if self.capacidadDisponible() >= peso / 1024:  
            self.canciones.append([nombre, artista, peso])
            print(f" Canción '{nombre}' añadida.")
        else:
            print(" No hay suficiente espacio para la canción.")
        return self

    def __sub__(self, video: list):
        nombre, artista, peso = video
        for v in self.videos:
            if v[0] == nombre and v[1] == artista:
                print(f"El video '{nombre}' ya existe.")
                return self

        if self.capacidadDisponible() >= peso:
            self.videos.append([nombre, artista, peso])
            print(f" Video '{nombre}' añadido.")
        else:
            print(" No hay suficiente espacio para el video.")
        return self

    def capacidadDisponible(self):
        total_usado_mb = 0
        for c in self.canciones:
            total_usado_mb += c[2] / 1024 
        for v in self.videos:
            total_usado_mb += v[2]    
        return self.capacidadGb * 1024 - total_usado_mb  

    def mostrarContenido(self):
        print(f"\nMP4 Marca: {self.marca}")
        print(f"Capacidad total: {self.capacidadGb} GB")
        print(f"Capacidad disponible: {self.capacidadDisponible():.2f} MB")

        print("\n Canciones:")
        for c in self.canciones:
            print(f"  - {c[0]} | {c[1]} | {c[2]} KB")

        print("\n Videos:")
        for v in self.videos:
            print(f"  - {v[0]} | {v[1]} | {v[2]} MB")

    def borrarCancion(self, nombre=None, artista=None, peso=None):
        originales = len(self.canciones)
        nuevas = []

        for c in self.canciones:
            if nombre and artista and peso is not None:
                if c[0] == nombre and c[1] == artista and c[2] == peso:
                    continue
            elif nombre and artista:
                if c[0] == nombre and c[1] == artista:
                    continue
            elif nombre:
                if c[0] == nombre:
                    continue
            elif artista:
                if c[1] == artista:
                    continue
            nuevas.append(c)

        eliminadas = originales - len(nuevas)
        self.canciones = nuevas
        print(f" Canciones eliminadas: {eliminadas}")

File: repos_estudiantes_ALVAREZ_primera_entrega/CARDON_PAZ_LUIS_ABDEL/POLIMORFISMO/EJERCICIO_13/MP4.java
package POLIMORFISMO.EJERCICIO_13;

public class MP4 {
    private String marca;
    private double capacidadGb;
    private int nroCanciones;
    private String[][] canciones;
    private int nroVideos;
    private String[][] videos;
    
    public MP4() {
        this.marca = "";
        this.capacidadGb = 0;
        this.nroCanciones = 0;
        this.nroVideos = 0;
        this.canciones = new String[100][3];
        this.videos = new String[100][3];
    }
    
    public MP4(String marca, double capacidadGb) {
        this.marca = marca;
        this.capacidadGb = capacidadGb;
        this.nroCanciones = 0;
        this.nroVideos = 0;
        this.canciones = new String[100][3];
        this.videos = new String[100][3];
    }
    
    public void agregarCancion(String nombre, String artista, double pesoKb) {
        if (nroCanciones < 100) {
            canciones[nroCanciones][0] = nombre;
            canciones[nroCanciones][1] = artista;
            canciones[nroCanciones][2] = String.valueOf(pesoKb);
            nroCanciones++;
        }
    }
    
    public void agregarVideo(String nombre, String artista, double pesoMb) {
        if (nroVideos < 100) {
            videos[nroVideos][0] = nombre;
            videos[nroVideos][1] = artista;
            videos[nroVideos][2] = String.valueOf(pesoMb);
            nroVideos++;
        }
    }

    public void borrar(String nombre) {
        for (int i = 0; i < nroCanciones; i++) {
            if (canciones[i][0] != null && canciones[i][0].equals(nombre)) {
                eliminarCancion(i);
                System.out.println("Canción '" + nombre + "' eliminada por nombre");
                return;
            }
        }
        System.out.println("Canción '" + nombre + "' no encontrada");
    }

    public void borrar(String artista, boolean porArtista) {
        int eliminadas = 0;
        for (int i = nroCanciones - 1; i >= 0; i--) {
            if (canciones[i][1] != null && canciones[i][1].equals(artista)) {
                eliminarCancion(i);
                eliminadas++;
            }
        }
        System.out.println(eliminadas + " canciones del artista '" + artista + "' eliminadas");
    }

    public void borrar(String nombre, double peso) {
        for (int i = 0; i < nroCanciones; i++) {
            if (canciones[i][0] != null && canciones[i][0].equals(nombre) &&
                Double.parseDouble(canciones[i][2]) == peso) {
                eliminarCancion(i);
                System.out.println("Canción '" + nombre + "' con peso " + peso + "Kb eliminada");
                return;
            }
        }
        System.out.println("Canción '" + nombre + "' con peso " + peso + "Kb no encontrada");
    }

    public boolean operatorPlus(String[] cancion) {
        if (nroCanciones >= 100) {
            System.out.println("No hay espacio para más canciones");
            return false;
        }

        for (int i = 0; i < nroCanciones; i++) {
            if (canciones[i][0] != null && canciones[i][0].equals(cancion[0]) &&
                canciones[i][1] != null && canciones[i][1].equals(cancion[1])) {
                System.out.println("La canción ya existe en el reproductor");
                return false;
            }
        }

        double pesoCancionKb = Double.parseDouble(cancion[2]);
        double capacidadUsada = calcularCapacidadUsada();
        double capacidadDisponible = capacidadGb * 1024 - capacidadUsada; 
        
        if (pesoCancionKb > capacidadDisponible) {
            System.out.println("No hay suficiente espacio para agregar la canción");
            return false;
        }

        canciones[nroCanciones][0] = cancion[0];
        canciones[nroCanciones][1] = cancion[1];
        canciones[nroCanciones][2] = cancion[2];
        nroCanciones++;
        System.out.println("Canción '" + cancion[0] + "' agregada exitosamente");
        return true;
    }

    public boolean operatorMinus(String[] video) {
        if (nroVideos >= 100) {
            System.out.println("No hay espacio para más videos");
            return false;
        }

        for (int i = 0; i < nroVideos; i++) {
            if (videos[i][0] != null && videos[i][0].equals(video[0]) &&
                videos[i][1] != null && videos[i][1].equals(video[1])) {
                System.out.println("El video ya existe en el reproductor");
                return false;
            }
        }

        double pesoVideoMb = Double.parseDouble(video[2]);
        double capacidadUsada = calcularCapacidadUsada();
        double capacidadDisponible = capacidadGb * 1024 - capacidadUsada; // Convertir GB a MB
        
        if (pesoVideoMb > capacidadDisponible) {
            System.out.println("No hay suficiente espacio para agregar el video");
            return false;
        }

        videos[nroVideos][0] = video[0];
        videos[nroVideos][1] = video[1];
        videos[nroVideos][2] = video[2];
        nroVideos++;
        System.out.println("Video '" + video[0] + "' agregado exitosamente");
        return true;
    }

    public void mostrarCapacidad() {
        double capacidadUsada = calcularCapacidadUsada();
        double capacidadDisponible = capacidadGb * 1024 - capacidadUsada;
        
        System.out.println("Capacidad total: " + capacidadGb + " GB");
        System.out.println("Capacidad usada: " + String.format("%.2f", capacidadUsada) + " MB");
        System.out.println("Capacidad disponible: " + String.format("%.2f", capacidadDisponible) + " MB");
        System.out.println("Canciones: " + nroCanciones);
        System.out.println("Videos: " + nroVideos);
    }

    private void eliminarCancion(int index) {
        for (int i = index; i < nroCanciones - 1; i++) {
            canciones[i][0] = canciones[i + 1][0];
            canciones[i][1] = canciones[i + 1][1];
            canciones[i][2] = canciones[i + 1][2];
        }
        nroCanciones--;
    }
    
    private double calcularCapacidadUsada() {
        double total = 0;

        for (int i = 0; i < nroCanciones; i++) {
            if (canciones[i][2] != null) {
                total += Double.parseDouble(canciones[i][2]) / 1024;
            }
        }

        for (int i = 0; i < nroVideos; i++) {
            if (videos[i][2] != null) {
                total += Double.parseDouble(videos[i][2]);
            }
        }
        
        return total;
    }

    public String getMarca() { 
        return marca; 
    }

    public int getNroCanciones() { 
        return nroCanciones; 
    }

    public int getNroVideos() { 
        return nroVideos; 
    }
} 
